src/test/overlay/overlay_test.cpp

```diff
@@ -0,0 +1,1403 @@
+            avg_bcache += node->overlay_->peerFinder().bootcacheSize();
+            bcache.push_back(node->overlay_->peerFinder().bootcacheSize());
+        }
+        auto stats = [](auto& avg, auto N, auto sample) {
+            avg = avg / N;
+            float sd = 0.;
+            for (auto d : sample)
+                sd += (d - avg) * (d - avg);
+            if (N > 1)
+                sd = sqrt(sd) / (N - 1);
+            return sd;
+        };
+        auto sd_peers_out = stats(avg_peers_out, Nout, peers_out);
+        auto sd_peers_in = stats(avg_peers_in, Nin, peers_in);
+        auto sd_pct_out = stats(avg_pct_out, Nout, pct_out);
+        auto sd_pct_in = stats(avg_pct_in, Nin, pct_in);
+        auto sd_pct_def_out =
+            stats(avg_pct_def_out, pct_def_out.size(), pct_def_out);
+        auto sd_pct_def_in =
+            stats(avg_pct_def_in, pct_def_in.size(), pct_def_in);
+        auto sd_lcache = stats(avg_lcache, lcache.size(), lcache);
+        auto sd_bcache = stats(avg_bcache, bcache.size(), bcache);
+        {
+            std::cout.precision(2);
+            std::cout << timeSinceStart() << ", out: " << tot_out
+                      << ", in: " << tot_in << ", snd: " << Counts::msgSendCnt
+                      << ", rcv: " << Counts::msgRecvCnt
+                      << ", deact: " << Counts::deactivateCnt
+                      << ", max out/in: " << out_max << "/" << in_max
+                      << ", avg out/in: " << avg_peers_out << "/"
+                      << sd_peers_out << ", " << avg_peers_in << "/"
+                      << sd_peers_in << ", avg pct out/in: " << avg_pct_out
+                      << "/" << sd_pct_out << ", " << avg_pct_in << "/"
+                      << sd_pct_in
+                      << ", avg pct default out/in: " << avg_pct_def_out << "/"
+                      << sd_pct_def_out << ", " << avg_pct_def_in << "/"
+                      << sd_pct_def_in << ", no peers: " << no_peers
+                      << ", live cache: " << avg_lcache << "/" << sd_lcache
+                      << ", boot cache: " << avg_bcache << "/" << sd_bcache
+                      << std::endl
+                      << std::flush;
+        }
+        if (tot_peers_in_.size() > 0 &&
+            tot_peers_in_[tot_peers_in_.size() - 1] != tot_in)
+            tot_peers_in_.clear();
+        if (tot_peers_out_.size() > 0 &&
+            tot_peers_out_[tot_peers_out_.size() - 1] != tot_out)
+            tot_peers_out_.clear();
+        tot_peers_in_.push_back(tot_in);
+        tot_peers_out_.push_back(tot_out);
+        // stop if the network doesn't change
+        if (tot_peers_in_.size() >= 6 && tot_peers_out_.size() >= 6)
+        {
+            return false;
+        }
+        return true;
+    }
+
+    void
+    run() override
+    {
+        testXRPLOverlay();
+    }
+};
+
+BEAST_DEFINE_TESTSUITE_MANUAL(overlay_net, overlay, ripple);
+BEAST_DEFINE_TESTSUITE_MANUAL(overlay_xrpl, overlay, ripple);
```

Comment:
These two tests fail for me:
```
Jade-5:~/Development/build> rippled --unittest=overlay_net
ripple.overlay.overlay_net Overlay
ERR:Server0 Bind port '' failed:Can't assign requested address             
#1 failed: unhandled exception: std::exception
313ms, 1 suite, 1 case, 1 test total, 1 failure
Jade-5:~/Development/build> rippled --unittest=overlay_xrpl
ripple.overlay.overlay_xrpl XRPLOverlay
#1 failed: adjacency matrix must be provided
102ms, 1 suite, 1 case, 1 test total, 1 failure
```

Reply:
Ip addresses have to be pre-configured for both tests. I have comments that explain how to run the test 
 and configure IP's: https://github.com/gregtatcam/rippled/blob/overlay-restr-partial1-templ/src/test/overlay/overlay_test.cpp#L63 and https://github.com/gregtatcam/rippled/blob/overlay-restr-partial1-templ/src/test/overlay/overlay_test.cpp#L1029. I can test for the configured IP's and provide a better error message if not configured and a better message if the adjacency matrix is missing. Does this work?

Reply:
Your response is good enough.  I didn't read the comments.  The way you have it sounds good to me, thanks.